- Description: Fixed dasd IO for eckd and fba with blocksize!=4k and mem > 2G
  Symptom:     IO errors reported in syslog, cannot boot with >2G memory
  Problem:     Calculation of number of idal words needed for channel program
               was incorrect.
  Solution:    Fix above calculation.
  Problem-ID:  
  Archs:       s390-64

--- linux-2.5/drivers/s390/block/dasd_fba.c	31 Mar 2004 15:04:28 -0000	1.33
+++ linux-2.5/drivers/s390/block/dasd_fba.c	17 Jun 2004 19:03:06 -0000	1.33.2.1
@@ -4,7 +4,7 @@
  * Bugreports.to..: <Linux390@de.ibm.com>
  * (C) IBM Corporation, IBM Deutschland Entwicklung GmbH, 1999,2000
  *
- * $Revision: 1.33 $
+ * $Revision: 1.33.2.1 $
  */
 
 #include <linux/config.h>
@@ -270,8 +270,8 @@
 				return ERR_PTR(-EINVAL);
 			count += bv->bv_len >> (device->s2b_shift + 9);
 #if defined(CONFIG_ARCH_S390X)
-			cidaw += idal_nr_words(page_address(bv->bv_page) +
-					       bv->bv_offset, bv->bv_len);
+			if (idal_is_needed (page_address(bv->bv_page), bv->bv_len))
+				cidaw += bv->bv_len / blksize;
 #endif
 		}
 	}

--- linux-2.5/drivers/s390/block/dasd_eckd.c	10 May 2004 12:46:18 -0000	1.53.2.2
+++ linux-2.5/drivers/s390/block/dasd_eckd.c	17 Jun 2004 19:03:06 -0000	1.53.2.3
@@ -7,7 +7,7 @@
  * Bugreports.to..: <Linux390@de.ibm.com>
  * (C) IBM Corporation, IBM Deutschland Entwicklung GmbH, 1999,2000
  *
- * $Revision: 1.53.2.2 $
+ * $Revision: 1.53.2.3 $
  */
 
 #include <linux/config.h>
@@ -983,8 +983,8 @@
 				return ERR_PTR(-EINVAL);
 			count += bv->bv_len >> (device->s2b_shift + 9);
 #if defined(CONFIG_ARCH_S390X)
-			cidaw += idal_nr_words(page_address(bv->bv_page) +
-					       bv->bv_offset, bv->bv_len);
+			if (idal_is_needed (page_address(bv->bv_page), bv->bv_len))
+				cidaw += bv->bv_len >> (device->s2b_shift + 9);
 #endif
 		}
 	}

